src/core/featurehistory.cpp

```diff
@@ -0,0 +1,448 @@
+    }
+
+    for ( const OldNewFeaturePair &pair : modifications.createdFeatures )
+    {
+      reversedModifications.deletedFeatures.append( OldNewFeaturePair( pair.second, pair.first ) );
+    }
+
+    for ( const OldNewFeaturePair &pair : modifications.updatedFeatures )
+    {
+      reversedModifications.updatedFeatures.append( OldNewFeaturePair( pair.second, pair.first ) );
+    }
+
+    reversedModificationsByLayerId.insert( layerId, reversedModifications );
+  }
+
+  return reversedModificationsByLayerId;
+}
+
+
+bool FeatureHistory::applyModifications( QMap<QString, FeatureModifications> &modificationsByLayerId )
+{
+  mIsApplyingModifications = true;
+
+  const QStringList layerIds = modificationsByLayerId.keys();
+
+  for ( const QString &layerId : layerIds )
+  {
+    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mProject->mapLayer( layerId ) );
+
+    if ( !vl )
+    {
+      continue;
+    }
+
+    FeatureModifications undoFeatureModifications = modificationsByLayerId.value( layerId );
+
+    if ( !vl->startEditing() )
+    {
+      return false;
+    }
+
+    // created features
+    QgsFeatureIds fidsToDelete;
+    for ( const OldNewFeaturePair &pair : undoFeatureModifications.createdFeatures )
+    {
+      fidsToDelete << pair.second.id();
+    }
+
+    if ( !undoFeatureModifications.createdFeatures.isEmpty() && !vl->deleteFeatures( fidsToDelete ) )
+    {
+      QgsMessageLog::logMessage( tr( "Failed to undo created features in layer \"%1\"" ).arg( vl->name() ) );
+      return false;
+    }
+
+    // deleted features
+    QgsFeatureList featuresToAdd;
+    for ( const OldNewFeaturePair &pair : undoFeatureModifications.deletedFeatures )
+    {
+      featuresToAdd.append( pair.first );
+    }
+
+    if ( !undoFeatureModifications.deletedFeatures.isEmpty() && !vl->addFeatures( featuresToAdd ) )
+    {
+      QgsMessageLog::logMessage( tr( "Failed to undo deleted features in layer \"%1\"" ).arg( vl->name() ) );
+      return false;
+    }
+
+    // update features
+    QgsFeatureIds featureIds;
+    for ( const OldNewFeaturePair &pair : undoFeatureModifications.updatedFeatures )
+    {
+      featureIds.insert( pair.first.id() );
+    }
+
+    for ( OldNewFeaturePair &pair : undoFeatureModifications.updatedFeatures )
+    {
+      if ( !vl->updateFeature( pair.first, true ) )
+      {
+        QgsMessageLog::logMessage( tr( "Failed to undo update features in layer \"%1\"" ).arg( vl->name() ) );
+        return false;
+      }
+    }
+
+    if ( !vl->commitChanges() )
+    {
+      QgsMessageLog::logMessage( tr( "Failed to commit undo feature modification in layer \"%1\"" ).arg( vl->name() ) );
+
+      if ( !vl->rollBack() )
+      {
+        QgsMessageLog::logMessage( tr( "Failed to rollback undo featurue modifications in layer \"%1\"" ).arg( vl->name() ) );
+      }
+
+      // OGR error creating feature -4: failed to execute insert : UNIQUE constraint failed: area.fid
```

Comment:
What's that?

Reply:
You just found one of the error messages that I had to debug :) . Deleted the line.